docs/proposals/self-hosted-kubelet.md

```diff
@@ -0,0 +1,75 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Proposal: Self-hosted kubelet
+
+## Abstract
+
+In a self-hosted Kubernetes deployment, we have the initial bootstrap problem. This proposal presents a solution to the kubelet bootstrap, and assumes a functioning control plane, and a kubelet that can securely contact the API server.
```

Comment:
In terms of the abstract, maybe we can elaborate a bit. Possibly something along the lines of:

> When running self-hosted components, there needs to be a mechanism for pivoting from the initial bootstrap state to the kubernetes-managed (self-hosted) state. In the case of a self-hosted kubelet, this means pivoting from the initial kubelet defined & run on the host, to the kubelet pod which has been scheduled to the node.

> This proposal presents a solution to the initial kubelet bootstrap, and the mechanism for pivoting to the self-hosted kubelet. This proposal assumes that the initial kubelet on the host is able to connect to a properly configured api-server.

---

Not sure if the above changes would answer the questions, but:

1) The bootstrap problem is essentially that we want the kubelet to be managed by kubernetes, but we need an initial kubelet to do that. So we need some mechanism for us to launch a kubelet, then give up control once a new kubelet has started.

2) A functioning apiserver would be required from the beginning (assuming no checkpointed pods I guess). Otherwise the initial kubelet would behave like any other kubelet without apiserver access.